### PR TITLE
[WIP] feat: initialize recipe for gperftools

### PR DESCRIPTION
**Pull request by vvarma**
Specify library name and version:  **gperftools/2.9.1**

This is also a good place to share with all of us **why you are submitting this PR** (specially if it is a new addition to ConanCenter): is it a dependency of other libraries you want to package? Are you the author of the library? Thanks!

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/gperftools/all/conandata.yml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+sources:
```
Comment (madebr) [inline recipes/gperftools/all/conandata.yml:1]: You need to add a `recipes/gperftools/config.yml` file that acts as an index.
See e.g. https://github.com/conan-io/conan-center-index/blob/master/recipes/7zip/config.yml

Comment (vvarma) [reply recipes/gperftools/all/conandata.yml:1]: thanks @madebr

File: recipes/gperftools/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        self.cpp_info.components["tcmalloc_and_profiler"].names["cmake_find_package_multi"] = "tcmalloc_and_profiler"
+        self.cpp_info.components["tcmalloc_and_profiler"].libs = ["tcmalloc_and_profiler"]
+
+        self.cpp_info.components["tcmalloc"].names["cmake_find_package"] = "tcmalloc"
+        self.cpp_info.components["tcmalloc"].names["cmake_find_package_multi"] = "tcmalloc"
+        self.cpp_info.components["tcmalloc"].libs = ["tcmalloc"]
+
+        self.cpp_info.components["profiler"].names["cmake_find_package"] = "profiler"
+        self.cpp_info.components["profiler"].names["cmake_find_package_multi"] = "profiler"
+        self.cpp_info.components["profiler"].libs = ["profiler"]
```
Comment (SSE4) [inline recipes/gperftools/all/conanfile.py:67]:
```suggestion
        self.cpp_info.components["profiler"].libs = ["profiler"]
        if self.settings.os == 'Linux':
            self.cpp_info.components['tcmalloc_and_profiler'].system_libs.extend(['pthread'])
```

File: recipes/gperftools/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+class GperftoolsConan(ConanFile):
+    name = "gperftools"
+    license = "BSD-3-Clause License"
+    homepage = "https://github.com/gperftools/gperftools"
+    url = "https://github.com/conan-io/conan-center-index"
+    description = "gperftools: originally Google Performance Tools"
+    topics = ("malloc", "profiler")
+    settings = "os", "compiler", "build_type", "arch"
+    options = {"shared": [True, False], "fPIC": [True, False]}
+    default_options = {"fPIC": False, "shared": True}
```
Comment (madebr) [inline recipes/gperftools/all/conanfile.py:14]: I am paid by the line count :smile: 
```suggestion
    options = {
        "shared": [True, False],
        "fPIC": [True, False],
    }
    default_options = {
        "shared": True,
        "fPIC": False,
    }
```

File: recipes/gperftools/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def source(self):
```
Comment (madebr) [inline recipes/gperftools/all/conanfile.py:25]: Remove fPIC when shared + you need libtool to run autoreconf.
```suggestion
            del self.options.fPIC

    def configure(self):
        if self.options.shared:
            del self.options.fPIC

    def build_requirements(self):
        self.build_requires("libtool/2.4.6")

    def source(self):
```

File: recipes/gperftools/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
+import os
+from conans import ConanFile, AutoToolsBuildEnvironment, tools
+
+
+class GperftoolsConan(ConanFile):
```
Comment (madebr) [inline recipes/gperftools/all/conanfile.py:5]:
```suggestion
from conans import ConanFile, AutoToolsBuildEnvironment, tools
import functools
import os

from conans import ConanInvalidConfiguration


class GperftoolsConan(ConanFile):
```

File: recipes/gperftools/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    name = "gperftools"
+    license = "BSD-3-Clause License"
+    homepage = "https://github.com/gperftools/gperftools"
+    url = "https://github.com/conan-io/conan-center-index"
+    description = "gperftools: originally Google Performance Tools"
+    topics = ("malloc", "profiler")
+    settings = "os", "compiler", "build_type", "arch"
+    options = {"shared": [True, False], "fPIC": [True, False]}
+    default_options = {"fPIC": False, "shared": True}
+    autotools = None
```
Comment (madebr) [inline recipes/gperftools/all/conanfile.py:15]:
```suggestion
```

File: recipes/gperftools/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+            else:
+                default_args.append("--disable-shared")
+            if self.options.fPIC:
+                default_args.append("--with-pic=yes")
+            autotools.configure(
+                configure_dir=self._source_subfolder,
+                args=default_args
+            )
+            self.autotools = autotools
+        return self.autotools
```
Comment (madebr) [inline recipes/gperftools/all/conanfile.py:46]: fPIC is handled automatically by the conan client.
But to be sure that static libraries are compiled with fpic when compiled as a shared library, set it explicitly.
Also, only run autoreconf only in build, before `_configure_autotools`.
```suggestion
    @functools.lru_cache(1)
    def _configure_autotools(self):
        autotools = AutoToolsBuildEnvironment(self, win_bash=tools.os_info.is_windows)
        yes_no = lambda v: "yes" if v else "no"
        args = [
            f"--enable-shared={yes_no(self.options.shared)}",
            f"--enable-static={yes_no(not self.options.shared)}",
        ]
        autotools.configure(args=args, configure_dir=self._source_subfolder)
        autotools.fpic = self.options.get_safe("fPIC", True)
        return autotools
```

File: recipes/gperftools/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+    def build(self):
+        autotools = self._configure_autotools()
```
Comment (madebr) [inline recipes/gperftools/all/conanfile.py:49]:
```suggestion
        with tools.chdir(self._source_subfolder):
            self.run("{} -fiv".format(tools.get_env("AUTORECONF"), win_bash=tools.os_info.is_windows)
        autotools = self._configure_autotools()
```

File: recipes/gperftools/all/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version],
+                  destination=self._source_subfolder, strip_root=True)
+
+    def _configure_autotools(self):
```
Comment (madebr) [inline recipes/gperftools/all/conanfile.py:29]: gperftools [provides a cmake script](https://github.com/gperftools/gperftools/blob/master/CMakeLists.txt),
why not use that?

File: recipes/gperftools/all/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+    def build(self):
+        autotools = self._configure_autotools()
+        autotools.make()
+
+    def package(self):
+        self.copy("COPYING", dst="licenses", src=self._source_subfolder)
+        autotools = self._configure_autotools()
+        autotools.install()
+        tools.rmdir(os.path.join(self.package_folder, "lib", "pkgconfig"))
```
Comment (madebr) [inline recipes/gperftools/all/conanfile.py:56]: Please also set the names for the `pkg_config` generator in `package_info`.

File: recipes/gperftools/all/conanfile.py [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    url = "https://github.com/conan-io/conan-center-index"
+    description = "gperftools: originally Google Performance Tools"
+    topics = ("malloc", "profiler")
+    settings = "os", "compiler", "build_type", "arch"
+    options = {
+        "shared": [True, False],
+        "fPIC": [True, False],
+    }
+    default_options = {
+        "shared": True,
```
Comment (madebr) [inline recipes/gperftools/all/conanfile.py:18]:
```suggestion
        "shared": False,
```

File: recipes/gperftools/all/conanfile.py [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+
+        self.cpp_info.components["tcmalloc"].names["cmake_find_package"] = "tcmalloc"
+        self.cpp_info.components["tcmalloc"].names["cmake_find_package_multi"] = "tcmalloc"
+        self.cpp_info.components["tcmalloc"].libs = ["tcmalloc"]
+
+        self.cpp_info.components["profiler"].names["cmake_find_package"] = "profiler"
+        self.cpp_info.components["profiler"].names["cmake_find_package_multi"] = "profiler"
+        self.cpp_info.components["profiler"].libs = ["profiler"]
+        if self.settings.os == 'Linux':
+            self.cpp_info.components['tcmalloc_and_profiler'].system_libs.extend(['pthread'])
```
Comment (madebr) [inline recipes/gperftools/all/conanfile.py:86]:
```suggestion
        if self.settings.os == "Linux":
            self.cpp_info.components["tcmalloc_and_profiler"].system_libs.append("pthread")
```

File: recipes/gperftools/all/conanfile.py (hunk 9 above)
Comment (vvarma) [reply recipes/gperftools/all/conanfile.py:45]: I was following the build instructions from the gperftools [INSTALL](https://github.com/gperftools/gperftools/blob/master/INSTALL) page. Also the cmake file has the following warning 
```cmake support is very preliminary. Autotools-based build is the only fully supported build for now.```

File: recipes/gperftools/all/conanfile.py [hunk 13]
```diff
@@ -0,0 +1,84 @@
+from conans import ConanFile, AutoToolsBuildEnvironment, tools
+import functools
+import os
+
+from conans import ConanInvalidConfiguration
```
Comment (madebr) [inline recipes/gperftools/all/conanfile.py:5]:
```suggestion
from conans import ConanFile, AutoToolsBuildEnvironment, tools
import functools
import os

required_conan_version = ">=1.33.0"
```

File: recipes/gperftools/all/conanfile.py [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+            f"--enable-shared={yes_no(self.options.shared)}",
+            f"--enable-static={yes_no(not self.options.shared)}",
+        ]
+        autotools.configure(args=args, configure_dir=self._source_subfolder)
+        autotools.fpic = self.options.get_safe("fPIC", True)
+        return autotools
+
+    def build(self):
+        with tools.chdir(self._source_subfolder):
+            self.run("{} -fiv".format(tools.get_env("AUTORECONF"), win_bash=tools.os_info.is_windows)
```
Comment (madebr) [inline recipes/gperftools/all/conanfile.py:58]:
```suggestion
            self.run("{} -fiv".format(tools.get_env("AUTORECONF")), win_bash=tools.os_info.is_windows)
```

File: recipes/gperftools/all/conanfile.py [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+    default_options = {
+        "shared": True,
+        "fPIC": False,
+    }
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    def config_options(self):
```
Comment (madebr) [inline recipes/gperftools/all/conanfile.py:29]:
```suggestion
        return "source_subfolder"

    @property
    def _settings_build(self):
        return getattr(self, "settings_build", self.settings)

    def config_options(self):
```

File: recipes/gperftools/all/conanfile.py [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
+
+    def build_requirements(self):
+        self.build_requires("libtool/2.4.6")
```
Comment (madebr) [inline recipes/gperftools/all/conanfile.py:38]:
```suggestion
        if self._settings_build.os == "Windows":
            self.build_requires("msys2/cci.latest")
        self.build_requires("libtool/2.4.6")
```

File: recipes/gperftools/all/conanfile.py (hunk 9 above)
Comment (madebr) [reply recipes/gperftools/all/conanfile.py:45]: Okay. autotools will probably be the most tested and guaranteed to have the most configuration options.

File: recipes/gperftools/all/conanfile.py (hunk 14 above)
Comment (madebr) [reply recipes/gperftools/all/conanfile.py:58]: This suggestion will fix the SyntaxError the ci is throwing atm.

**Comment by vvarma** [comment]
@madebr I am still testing the usability of this library. Can I mark it as draft somehow?

File: recipes/gperftools/all/conanfile.py (hunk 14 above)
Comment (madebr) [reply recipes/gperftools/all/conanfile.py:58]: @vvarma 
Please apply this suggestion. CI currently fails because of a syntax error.

**Comment by madebr** [comment]
> @madebr I am still testing the usability of this library. Can I mark it as draft somehow?

Yes sure.
Draft pr's are still tested. Draft pr's can't just get merged.

**Comment by vvarma** [comment]
Closing this PR for now. Will reopen this when I have the time to fix it again.

**Pull request closed without merging**